archive/100-theorems-list/90_stirlings_formula.lean

```diff
@@ -0,0 +1,287 @@
+  norm_num,
+  simp only [one_div],
+  intro m,
+  refine mul_nonneg _ _,
+  all_goals {refine inv_nonneg.mpr _, norm_cast, exact (zero_le _)},
+end
+
+/--
+We have the bound  `log_stirling_seq n - log_stirling_seq (n+1) ≤ 1/(2n+1)^2* 1/(1-(1/2n+1)^2)`.
+-/
+lemma log_stirling_seq_diff_le_geo_sum : ∀ (n : ℕ),
+  log_stirling_seq n.succ - log_stirling_seq n.succ.succ ≤
+  (1 / (2 * n.succ + 1)) ^ 2 / (1 - (1 / (2 * n.succ + 1)) ^ 2) :=
+begin
+  intro n,
+  have h_nonneg : 0 ≤ ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2),
+  by { rw [cast_succ, one_div, inv_pow, inv_nonneg], norm_cast, exact zero_le', },
+  have g : has_sum (λ (k : ℕ), ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2) ^ k.succ)
+    ((1 / (2 * n.succ + 1)) ^ 2 / (1 - (1 / (2 * n.succ + 1)) ^ 2)) :=
+  begin
+    have h_pow_succ := λ (k : ℕ),
+      symm (pow_succ ((1 / (2 * ((n : ℝ) + 1) + 1)) ^ 2) k),
+    have hlt : ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2) < 1, by
+    { simp only [cast_succ, one_div, inv_pow],
+      refine inv_lt_one _,
+      norm_cast,
+      simp only [nat.one_lt_pow_iff, ne.def, zero_eq_bit0, nat.one_ne_zero, not_false_iff,
+        lt_add_iff_pos_left, canonically_ordered_comm_semiring.mul_pos, succ_pos', and_self], },
+    exact (has_sum_geometric_of_lt_1 h_nonneg hlt).mul_left ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2)
+  end,
+  have hab :
+    ∀ (k : ℕ), (1 / (2 * (k.succ : ℝ) + 1)) * ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2) ^ k.succ ≤
+    ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2) ^ k.succ :=
+  begin
+    intro k,
+    have h_zero_le : 0 ≤ ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2) ^ k.succ := pow_nonneg h_nonneg _,
+    have h_left : 1 / (2 * (k.succ : ℝ) + 1) ≤ 1, by
+    { simp only [cast_succ, one_div],
+      refine inv_le_one _,
+      norm_cast,
+      exact (le_add_iff_nonneg_left 1).mpr zero_le', },
+    exact mul_le_of_le_one_left h_zero_le h_left,
+  end,
+  exact has_sum_le hab (log_stirling_seq_diff_has_sum n) g,
+end
+
+/--
+We have the bound  `log_stirling_seq n - log_stirling_seq (n+1)` ≤ 1/(4 n^2)
+-/
+lemma log_stirling_seq_sub_log_stirling_seq_succ (n : ℕ) :
+  log_stirling_seq n.succ - log_stirling_seq n.succ.succ ≤ 1 / (4 * n.succ ^ 2) :=
+begin
+  have h₁ : 0 < 4 * ((n:ℝ) + 1)^2 := by nlinarith [@cast_nonneg ℝ _ n],
+  have h₃ : 0 < (2 * ((n:ℝ) + 1) + 1) ^ 2 := by nlinarith [@cast_nonneg ℝ _ n ],
+  have h₂ : 0 < 1 - (1 / (2 * ((n:ℝ) + 1) + 1)) ^ 2,
+  { rw ← mul_lt_mul_right h₃,
+    have H : 0 < (2 * ((n:ℝ) + 1) + 1) ^ 2 - 1 := by nlinarith [@cast_nonneg ℝ _ n ],
+    convert H using 1; field_simp [h₃.ne'] },
+  refine le_trans (log_stirling_seq_diff_le_geo_sum n) _,
+  push_cast at *,
+  rw div_le_div_iff h₂ h₁,
+  field_simp [h₃.ne'],
+  rw div_le_div_right h₃,
+  ring_nf,
+  norm_cast,
+  linarith,
+end
+
+/-- For any `n`, we have `log_stirling_seq 1 - log_stirling_seq n ≤ 1/4 * ∑' 1/k^2`  -/
+lemma log_stirling_seq_bounded_aux : ∃ (c : ℝ), ∀ (n : ℕ),
+log_stirling_seq 1 - log_stirling_seq n.succ ≤ c :=
+begin
+  let d := ∑' k : ℕ, (1 : ℝ) / (k.succ)^2,
+  use (1/4 * d : ℝ),
```

Comment:
```suggestion
  use (1 / 4 * d : ℝ),
```
and same in a number of other places. This is just a personal style nitpick though, so feel free to ignore.

Reply:
That said... is it not more convenient to use `d / 4` instead?

Reply:
`1 / 4 * d` is the first thing that drops out after removing dependency from `n`, so having `d / 4` would mean an additional line in the calc proof below.